File: frontend/src/customer/object/profile/Coupon.ts
import type {
  CouponId,
  CurrencyCents,
  DescriptionText,
  IsoDateTime,
} from '@/shared/object/domain/DomainObjects'

export type Coupon = {
  id: CouponId
  title: DescriptionText
  discountCents: CurrencyCents
  minimumSpendCents: CurrencyCents
  description: DescriptionText
  expiresAt: IsoDateTime
}

File: backend/src/main/scala/domain/customer/Coupon.scala
package domain.customer

import domain.shared.given

import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.*
import domain.shared.*

final case class Coupon(
    id: CouponId,
    title: DisplayText,
    discountCents: CurrencyCents,
    minimumSpendCents: CurrencyCents,
    description: DescriptionText,
    expiresAt: IsoDateTime,
)
object Coupon:
  given Encoder[Coupon] = deriveEncoder
  given Decoder[Coupon] = deriveDecoder
